open3d_slam/src/ScanToMapRegistration.cpp

```diff
@@ -0,0 +1,93 @@
+/*
+ * ScanToMapRegistration.cpp
+ *
+ *  Created on: Oct 31, 2022
+ *      Author: jelavice
+ */
+#include "open3d_slam/ScanToMapRegistration.hpp"
+#include "open3d_slam/Submap.hpp"
+#include "open3d_slam/helpers.hpp"
+#include "open3d_slam/assert.hpp"
+
+#include "open3d_slam/cloud_processing.hpp"
+
+namespace o3d_slam {
+
+namespace {
+namespace registration = open3d::pipelines::registration;
+std::shared_ptr<registration::TransformationEstimation> icpObjective;
+open3d::pipelines::registration::ICPConvergenceCriteria icpCriteria;
+} // namespace
+
+ScanToMapIcpOpen3D::ScanToMapIcpOpen3D() {
+	update(params_);
+}
+
+void ScanToMapIcpOpen3D::setParameters(const MapperParameters &p) {
+	params_ = p;
+	update(params_);
+}
+void ScanToMapIcpOpen3D::update(const MapperParameters &p) {
+	icpCriteria.max_iteration_ = p.scanMatcher_.maxNumIter_;
+	icpObjective = icpObjectiveFactory(p.scanMatcher_.icpObjective_);
+	mapBuilderCropper_ = croppingVolumeFactory(params_.mapBuilder_.cropper_);
+	scanMatcherCropper_ = croppingVolumeFactory(params_.scanProcessing_.cropper_);
+}
+
+void ScanToMapIcpOpen3D::estimateNormalsIfNeeded(PointCloud *pcl) const {
```

Comment:
Maybe this function can be moved to cloud_processing or with a possible rename: `cloudProcessor` 

Motivation: In the near future we might estimate more than just the normals i.e. curvature, lines, local variance
that way we can just keep adding to this processor class

Reply:
I added another abstraction for scan-to-scan matching. The scan2map refinement can be tricky if you change the underlying map representation (e.g. you move towards something with voxelized representation and/or voxel hasning, etc.). This is why I kept this the interface where you pass in the Submap class.

Not sure that is the best solution, but it's an improvement compared to what we had earlier where everything was depending on the particular ICP implementation.

Reply:
Thanks for the explanation! This separation is equally beneficial is you already have scan2scan abstraction then forget my previous suggestion :+1: